File: legacy_prototype/src/flowa/lexer.py
import re
from typing import List, Optional
from .tokens import Token, TokenType

class LexerError(Exception):
    def __init__(self, message, line, column):
        super().__init__(f"{message} at line {line}, column {column}")
        self.line = line
        self.column = column

class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
        self.indent_stack = [0]  # Stack of indentation levels (column numbers)
        
        self.keywords = {
            'def': TokenType.DEF,
            'async': TokenType.ASYNC,
            'return': TokenType.RETURN,
            'if': TokenType.IF,
            'else': TokenType.ELSE,
            'while': TokenType.WHILE,
            'for': TokenType.FOR,
            'in': TokenType.IN,
            'spawn': TokenType.SPAWN,
            'await': TokenType.AWAIT,
            'True': TokenType.TRUE,
            'False': TokenType.FALSE,
            'None': TokenType.NONE,
            'lambda': TokenType.LAMBDA,
        }

    def tokenize(self) -> List[Token]:
        while self.pos < len(self.source):
            char = self.peek()
            
            if char == '#':
                self.skip_comment()
            elif char == '\n':
                self.handle_newline()
            elif char.isspace():
                self.advance()
            elif char.isalpha() or char == '_':
                self.tokens.append(self.read_identifier())
            elif char.isdigit():
                self.tokens.append(self.read_number())
            elif char == '"' or char == "'":
                self.tokens.append(self.read_string())
            else:
                self.handle_symbol()
                
        self.handle_eof()
        return self.tokens

    def peek(self) -> str:
        if self.pos >= len(self.source):
            return ''
        return self.source[self.pos]
    
    def advance(self):
        if self.pos < len(self.source):
            if self.source[self.pos] == '\n':
                self.line += 1
                self.column = 1
            else:
                self.column += 1
            self.pos += 1

    def skip_comment(self):
        while self.peek() != '\n' and self.peek() != '':
            self.advance()

    def handle_newline(self):
        self.advance() # Consume \n
        
        # Check for empty lines or comment-only lines
        start_pos = self.pos
        while self.peek().isspace() and self.peek() != '\n' and self.peek() != '':
            self.advance()
            
        if self.peek() == '\n' or self.peek() == '#':
            return # Ignore empty lines
            
        if self.peek() == '':
            return # EOF after newline

        # Calculate indentation
        indent_level = 0
        # Reset to start of line to count spaces properly
        # Actually, since we advanced, column is reset to 1. 
        # But we need to count spaces from the start of the line.
        # Let's just count the spaces we skipped.
        # Wait, self.column is 1 after \n. 
        # The loop above advanced over spaces. 
        # So current column - 1 is the indentation level if we assume 1-based indexing and start at 1.
        indent_level = self.column - 1
        
        current_indent = self.indent_stack[-1]
        
        # Emit NEWLINE for the previous line
        self.tokens.append(Token(TokenType.NEWLINE, line=self.line-1, column=1))

        if indent_level > current_indent:
            self.indent_stack.append(indent_level)
            self.tokens.append(Token(TokenType.INDENT, line=self.line, column=1))
        elif indent_level < current_indent:
            while indent_level < self.indent_stack[-1]:
                self.indent_stack.pop()
                self.tokens.append(Token(TokenType.DEDENT, line=self.line, column=1))
            if indent_level != self.indent_stack[-1]:
                raise LexerError("Inconsistent indentation", self.line, 1)
        else:
            pass

    def read_identifier(self) -> Token:
        start_col = self.column
        value = ''
        while self.peek().isalnum() or self.peek() == '_':
            value += self.peek()
            self.advance()
            
        token_type = self.keywords.get(value, TokenType.IDENTIFIER)
        return Token(token_type, value, self.line, start_col)

    def read_number(self) -> Token:
        start_col = self.column
        value = ''
        is_float = False
        while self.peek().isdigit() or self.peek() == '.':
            if self.peek() == '.':
                if is_float: break
                is_float = True
            value += self.peek()
            self.advance()
            
        return Token(TokenType.NUMBER, float(value) if is_float else int(value), self.line, start_col)

    def read_string(self) -> Token:
        start_col = self.column
        quote = self.peek()
        self.advance()
        value = ''
        while self.peek() != quote and self.peek() != '':
            if self.peek() == '\\':
                self.advance()
                # Handle escapes if needed
            value += self.peek()
            self.advance()
        
        if self.peek() == quote:
            self.advance()
        else:
            raise LexerError("Unterminated string", self.line, self.column)
            
        return Token(TokenType.STRING, value, self.line, start_col)

    def handle_symbol(self):
        start_col = self.column
        char = self.peek()
        self.advance()
        
        token_type = None
        
        # Two-char operators
        if char == '|' and self.peek() == '>':
            self.advance()
            token_type = TokenType.PIPE
        elif char == '=' and self.peek() == '=':
            self.advance()
            token_type = TokenType.EQEQ
        elif char == '!' and self.peek() == '=':
            self.advance()
            token_type = TokenType.NEQ
        elif char == '<' and self.peek() == '=':
            self.advance()
            token_type = TokenType.LTE
        elif char == '>' and self.peek() == '=':
            self.advance()
            token_type = TokenType.GTE
        elif char == '-' and self.peek() == '>':
            self.advance()
            token_type = TokenType.ARROW
            
        # Single-char operators
        elif char == '+': token_type = TokenType.PLUS
        elif char == '-': token_type = TokenType.MINUS
        elif char == '*': token_type = TokenType.STAR
        elif char == '/': token_type = TokenType.SLASH
        elif char == '=': token_type = TokenType.EQ
        elif char == '<': token_type = TokenType.LT
        elif char == '>': token_type = TokenType.GT
        elif char == '(': token_type = TokenType.LPAREN
        elif char == ')': token_type = TokenType.RPAREN
        elif char == '{': token_type = TokenType.LBRACE
        elif char == '}': token_type = TokenType.RBRACE
        elif char == '[': token_type = TokenType.LBRACKET
        elif char == ']': token_type = TokenType.RBRACKET
        elif char == ',': token_type = TokenType.COMMA
        elif char == ':': token_type = TokenType.COLON
        elif char == '.': token_type = TokenType.DOT
        
        if token_type:
            self.tokens.append(Token(token_type, char, self.line, start_col))
        else:
            raise LexerError(f"Unexpected character: {char}", self.line, start_col)

    def handle_eof(self):
        # Emit NEWLINE if not just emitted?
        if self.tokens and self.tokens[-1].type != TokenType.NEWLINE and self.tokens[-1].type != TokenType.DEDENT:
             self.tokens.append(Token(TokenType.NEWLINE, line=self.line, column=1))
             
        # Dedent remaining
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            self.tokens.append(Token(TokenType.DEDENT, line=self.line, column=1))
            
        self.tokens.append(Token(TokenType.EOF, line=self.line, column=1))

File: pkg/lexer/lexer.go
package lexer

import (
	"flowa/pkg/token"
	"strings"
)

type Lexer struct {
	input        string
	position     int  // current position in input (points to current char)
	readPosition int  // current reading position in input (after current char)
	ch           byte // current char under examination
	line         int
	column       int

	indentStack []int // Stack of indentation levels (column numbers)
	tokenQueue  []token.Token
}

func New(input string) *Lexer {
	l := &Lexer{
		input:       input,
		line:        1,
		column:      0,
		indentStack: []int{0},
	}
	l.readChar()
	return l
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition += 1
	l.column += 1
}

func (l *Lexer) peekChar() byte {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

func (l *Lexer) NextToken() token.Token {
	// If we have queued tokens (e.g. DEDENTs), return them first
	if len(l.tokenQueue) > 0 {
		tok := l.tokenQueue[0]
		l.tokenQueue = l.tokenQueue[1:]
		return tok
	}

	var tok token.Token

	// Skip whitespace but NOT newlines
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\r' {
		l.readChar()
	}

	// Skip comments
	if l.ch == '#' {
		l.skipComment()
		// After skipping comment, we're at newline or EOF
		// Continue to next token
		return l.NextToken()
	}

	switch l.ch {
	case '\n':
		return l.handleNewline()
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.EQ, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.ASSIGN, l.ch, l.line, l.column)
		}
	case '+':
		if l.peekChar() == '+' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.PLUS_PLUS, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.PLUS, l.ch, l.line, l.column)
		}
	case '-':
		if l.peekChar() == '>' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.ARROW, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else if l.peekChar() == '-' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.MINUS_MINUS, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.MINUS, l.ch, l.line, l.column)
		}
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.NOT_EQ, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.BANG, l.ch, l.line, l.column)
		}
	case '/':
		tok = newToken(token.SLASH, l.ch, l.line, l.column)
	case '*':
		tok = newToken(token.ASTERISK, l.ch, l.line, l.column)
	case '<':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.LTE, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.LT, l.ch, l.line, l.column)
		}
	case '>':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.GTE, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.GT, l.ch, l.line, l.column)
		}
	case '|':
		if l.peekChar() == '>' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.PIPE, Literal: string(ch) + string(l.ch), Line: l.line, Column: l.column}
		} else {
			tok = newToken(token.ILLEGAL, l.ch, l.line, l.column)
		}
	case ',':
		tok = newToken(token.COMMA, l.ch, l.line, l.column)
	case ':':
		tok = newToken(token.COLON, l.ch, l.line, l.column)
	case ';':
		tok = newToken(token.SEMICOLON, l.ch, l.line, l.column)
	case '(':
		tok = newToken(token.LPAREN, l.ch, l.line, l.column)
	case ')':
		tok = newToken(token.RPAREN, l.ch, l.line, l.column)
	case '.':
		tok = newToken(token.DOT, l.ch, l.line, l.column)
	case '{':
		tok = newToken(token.LBRACE, l.ch, l.line, l.column)
	case '}':
		tok = newToken(token.RBRACE, l.ch, l.line, l.column)
	case '[':
		tok = newToken(token.LBRACKET, l.ch, l.line, l.column)
	case ']':
		tok = newToken(token.RBRACKET, l.ch, l.line, l.column)
	case '"':
		tok.Type = token.STRING
		tok.Literal = l.readString()
		tok.Line = l.line
		tok.Column = l.column
	case 0:
		// Handle EOF: dedent remaining
		if len(l.indentStack) > 1 {
			for i := 0; i < len(l.indentStack)-1; i++ {
				l.tokenQueue = append(l.tokenQueue, token.Token{Type: token.DEDENT, Literal: "", Line: l.line, Column: l.column})
			}
			l.indentStack = []int{0}
			tok = l.tokenQueue[0]
			l.tokenQueue = l.tokenQueue[1:]
		} else {
			tok.Literal = ""
			tok.Type = token.EOF
		}
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = token.LookupIdent(tok.Literal)
			// fmt.Printf("DEBUG: Ident=%q Type=%q\n", tok.Literal, tok.Type)
			tok.Line = l.line
			tok.Column = l.column
			return tok
		} else if isDigit(l.ch) {
			tok.Type = token.INT
			tok.Literal = l.readNumber()
			tok.Line = l.line
			tok.Column = l.column
			return tok
		} else {
			tok = newToken(token.ILLEGAL, l.ch, l.line, l.column)
		}
	}

	l.readChar()
	return tok
}

func (l *Lexer) handleNewline() token.Token {
	// 1. Emit NEWLINE
	tok := token.Token{Type: token.NEWLINE, Literal: "\n", Line: l.line, Column: l.column}

	l.readChar() // Consume \n
	l.line++
	l.column = 0 // Reset column for new line (will be 1 after next readChar if we didn't do it here, but we are consuming manually)
	// Actually readChar increments column. So if we just consumed \n, column is 1?
	// Let's reset column to 0, so next char is 1.
	l.column = 0

	// 2. Check indentation
	// Peek ahead to count spaces
	indentLen := 0
	for l.ch == ' ' || l.ch == '\t' {
		if l.ch == '\t' {
			indentLen += 4 // Assume tab is 4 spaces for now
		} else {
			indentLen++
		}
		l.readChar()
	}

	// If line is empty or comment, ignore indentation change
	if l.ch == '\n' || l.ch == 0 {
		// Just a blank line, maybe emit another newline or ignore?
		// Usually we ignore blank lines in indentation logic.
		// But we already emitted NEWLINE.
		// Let's just return the NEWLINE we created.
		return tok
	}

	currentIndent := l.indentStack[len(l.indentStack)-1]

	if indentLen > currentIndent {
		l.indentStack = append(l.indentStack, indentLen)
		l.tokenQueue = append(l.tokenQueue, token.Token{Type: token.INDENT, Literal: "", Line: l.line, Column: l.column})
	} else if indentLen < currentIndent {
		for len(l.indentStack) > 1 && indentLen < l.indentStack[len(l.indentStack)-1] {
			l.indentStack = l.indentStack[:len(l.indentStack)-1]
			l.tokenQueue = append(l.tokenQueue, token.Token{Type: token.DEDENT, Literal: "", Line: l.line, Column: l.column})
		}
		if indentLen != l.indentStack[len(l.indentStack)-1] {
			// Indentation error
			return token.Token{Type: token.ILLEGAL, Literal: "Indentation Error", Line: l.line, Column: l.column}
		}
	}

	return tok
}

func newToken(tokenType token.TokenType, ch byte, line, col int) token.Token {
	return token.Token{Type: tokenType, Literal: string(ch), Line: line, Column: col}
}

func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) || isDigit(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func (l *Lexer) readNumber() string {
	position := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	// Support float?
	if l.ch == '.' {
		l.readChar()
		for isDigit(l.ch) {
			l.readChar()
		}
	}
	return l.input[position:l.position]
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func (l *Lexer) readString() string {
	var result strings.Builder
	l.readChar() // Skip opening quote

	for l.ch != '"' && l.ch != 0 {
		if l.ch == '\\' {
			// Handle escape sequences
			l.readChar()
			switch l.ch {
			case 'n':
				result.WriteByte('\n')
			case 't':
				result.WriteByte('\t')
			case 'r':
				result.WriteByte('\r')
			case '\\':
				result.WriteByte('\\')
			case '"':
				result.WriteByte('"')
			case '0':
				result.WriteByte('\x00')
			default:
				// Unknown escape, just include the backslash and character
				result.WriteByte('\\')
				result.WriteByte(l.ch)
			}
		} else {
			result.WriteByte(l.ch)
		}
		l.readChar()
	}

	return result.String()
}

func (l *Lexer) skipComment() {
	for l.ch != '\n' && l.ch != 0 {
		l.readChar()
	}
}
